src/aignostics/application/_service.py

```diff
@@ -366,10 +449,17 @@ def generate_metadata_from_source_directory(  # noqa: PLR0913, PLR0917
 
         metadata = []
 
+        # Pre-filter: exclude redundant DICOM files from multi-file pyramids
+        dicom_files_to_exclude = Service._filter_dicom_pyramid_files(source_directory)
+
         try:
             extensions = get_supported_extensions_for_application(application_id)
             for extension in extensions:
                 for file_path in source_directory.glob(f"**/*{extension}"):
+                    # Skip excluded DICOM files
+                    if file_path in dicom_files_to_exclude:
+                        continue
```

Comment:
This is somewhat inefficient:
1. First we list all files in the directory and derive files to exclude
2. Then we list all files in the directory again, and skip files from (1)

Instead we could have a single function which lists files in the directory and returns only those to **include**, and iterate over that?